content/drafts/sphinx.md

```diff
@@ -0,0 +1,447 @@
+
+The diagram above aims to show how this can be done between two parties who own 
+keys `A` and `B` respectively. The private keys `a` and `b` remain private and 
+yet the two parties are able to derive a shared secret key. To prove to yourself 
+that this is indeed a shared secret key, ask yourself if a third party with 
+private key `q` and public key `Q` could derive the shared secret between `A` 
+and `B` if `Q` had knowledge of both the `A` and `B` public keys. You should be 
+able to see that the answer is no.
+
+In Sphinx packet construction, the onion packet creator, Alice, will use a 
+shared secret key between a public key that she owns and the public key of each 
+node on the path in order to encrypt data for the relevant hop. 
+
+Once a hop has derived the shared secret with Alice, they can then use that 
+shared secret along with some constants (`rho`, `mu`, `pad`, `um` and `ammag`) 
+to derive other keys or byte streams:
+
+![](/onion/4-key_derivation.png#center)
+
+# Ephemeral Keys
+
+Each hop on the path will need to be told what the public key of the sender is 
+so that they can derive the shared secret with the sender and decrypt the 
+packet. However, sharing the node level public key of the sender, `A`,  with 
+each hop is a bit of a privacy leak. Everyone on the path will know just how 
+much Alice is spending on coffee. To prevent this, Alice will instead use an 
+ephemeral key with each hop. This means that she will have a new private and 
+public key pair for each hop on the path and will use that pair to derive the 
+shared secret with the hop’s public key. After the payment is complete, Alice 
+can discard this ephemeral key pair.
+
+Naively, Alice could derive a completely new and random key pair for each hop 
+along the path but that means she would need to persist each private key while 
+the payment completes, _and_ she would need to include the ephemeral public key 
+for each hop in the onion packet for that hop. This uses up quite a bit of onion 
+packet space.
+
+Instead, what she will do is derive a single ephemeral key pair. The private key 
+of this pair is called the `session_key`. Alice uses this key to derive all the 
+other ephemeral keys along the path. What is cool about this is that she only 
+has to communicate the very first public ephemeral key to the first hop. That 
+hop will then be able to use that ephemeral key along with its shared secret 
+with Alice to derive the ephemeral key to pass on to the following hop. This 
+should become more clear with the long-running example later. 
+
+# HMAC
+
+A [Hash-Based Message Authentication Code (or HMAC)][hmac] is a message 
+authentication technique. Basically, a party with a private-public key pair, 
+can use the private key along with the hash of a message they would like to 
+sign to produce a 32 byte “signature” called an HMAC. Then any other party can 
+be given the message, the HMAC and the signer’s public key and can verify that 
```

Comment:
This is actually one key difference between a HMAC and ECDSA: To verify the HMAC, you need the secret/private key. You cannot verify an HMAC with just the public key. Because basically the "key" to the HMAC can be any data of any arbitrary length, so it doesn't have to be an EC private key at all.

Reply:
you're totally right!! 🙈  updating asap